src/abstractdataframe/selectionfast.jl

```diff
@@ -0,0 +1,180 @@
+`df_sel` argument is a data frame storing columns selected by `AsTable(...)`
+selector. By default it calls `default_table_transformation`. However, it is
+allowed to add special methods for specific types of `fun` with the reservation
+that the produced result must match the result that would be produced by
+`default_table_transformation`, except that it is allowed to perform `eltype`
+conversion of the resulting vectors or value type promotions that are consistent
+with `promote_type`.
+
+The main use of special `table_transformation` methods is to provide more
+efficient than the default implementations of requested `fun` transformation.
+"""
+@noinline table_transformation(df_sel::AbstractDataFrame, fun) =
+    default_table_transformation(df_sel, fun)
+
+"""
+    default_table_transformation(df_sel::AbstractDataFrame, fun)
+
+This is a default implementation called when `AsTable(...) => fun` is requested.
+The `df_sel` argument is a data frame storing columns selected by
+`AsTable(...)` selector.
+"""
+default_table_transformation(df_sel::AbstractDataFrame, fun) =
+    fun(Tables.columntable(df_sel))
+
+@noinline table_transformation(df_sel::AbstractDataFrame, ::typeof(sum)) =
+    _sum_fast(map(identity, eachcol(df_sel)))
+
+@noinline table_transformation(df_sel::AbstractDataFrame, ::ByRow{typeof(sum)}) =
+    table_transformation(df_sel, sum)
+
+function _sum_fast(cols::Vector{<:AbstractVector})
+    isempty(cols) && throw(ArgumentError("No columns selected for reduction"))
+    return sum(cols)
+end
+
+@noinline function table_transformation(df_sel::AbstractDataFrame,
+                                        fun::typeof(ByRow(sum∘skipmissing)))
+    fastsum = _sum_skipmissing_fast(map(identity, eachcol(df_sel)))
+    isnothing(fastsum) || return fastsum
+    slowsum = default_table_transformation(df_sel, fun)
+    isconcretetype(nonmissingtype(eltype(slowsum))) && return slowsum
+    T = mapreduce(typeof, promote_type, slowsum)
+    return convert(AbstractVector{T}, slowsum)
+end
+
+function _sum_skipmissing_fast(cols::Vector{<:AbstractVector})
+    @assert !isempty(cols)
+    local sumz
+    sumz_undefined = true
+    for col in cols
+        try
+            zi = zero(eltype(col))
+            if sumz_undefined
+                sumz_undefined = false
+                sumz = zi
+            elseif !ismissing(zi) # zi is missing if eltype is Missing
+                sumz += zi
+            end
+        catch e
+            if e isa MethodError && e.f === zero
+                sumz_undefined = true
+                break
+            else
+                throw(e)
+            end
+        end
+    end
+    # this will happen if eltype of some columns do not support zero
+    # or all columns have eltype Missing
+    sumz_undefined && return nothing
+    init = fill!(Tables.allocatecolumn(typeof(sumz), length(cols[1])), sumz)
+
+    return foldl(cols, init=init) do l, r
+        l .= ifelse.(ismissing.(r), l, l .+ r)
+    end
```

Comment:
Is `foldl` really better than a `for` loop here? Anyway, better use the same approach here and in `mean` below.

Reply:
I will change to loop